[BUG] Biometric unlock of database lose key file config

### Checks

- [x] I have read the FAQ section, searched the open issues, and still think this is a new bug.

### Describe the bug you encountered:

Since the last 1.12-r5 version installed from the beta program on play store, when I enable biometric full unlock for my database the app sometime forget to use the key file and the fingerprint unlock then fails.
I then have to select again the key file and enter my password.
I'm on a Pixel 7 with Android 15.

### Describe what you expected to happen:

_No response_

### What version of Keepass2Android are you using?

1.12-r5

### Which version of Android are you on?

15

Here are the logs of the error :

```
12/05/2025 23:55:25:786 -- CFS: OpenWhenNoLocalChanges
12/05/2025 23:55:25:787 -- CFS: hashing cached version
12/05/2025 23:55:25:843 -- PasswordModeSpinner item selected: 0
12/05/2025 23:55:25:843 -- PasswordModeSpinner item selected: 0
12/05/2025 23:55:26:347 -- SelectCurrentDbActivity.OnStop 1
12/05/2025 23:55:26:348 -- FileSelect.OnStop
12/05/2025 23:55:26:378 -- FileSelect.OnDestroyTrue
12/05/2025 23:55:26:379 -- SelectCurrentDbActivity.OnStop 2
12/05/2025 23:55:26:380 -- SelectCurrentDbActivity.OnDestroyTrue 2
12/05/2025 23:55:26:381 -- FileSelect.OnStop
12/05/2025 23:55:26:403 -- FileSelect.OnDestroyTrue
12/05/2025 23:55:27:636 -- FP: Decrypting 
12/05/2025 23:55:27:645 -- GeneralSecurityException in DecryptStored
12/05/2025 23:55:27:666 -- Java.Security.GeneralSecurityException: Exception_WasThrown, Java.Security.GeneralSecurityException
 ---> Java.Lang.Exception: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:835: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:385: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:90: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED))

  --- End of managed Java.Lang.Exception stack trace ---
android.security.KeyStoreException: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:835: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:385: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:90: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)) (public error code: 2 internal Keystore code: -26)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:428)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:79)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:118)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:179)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:624)
	at javax.crypto.Cipher.doFinal(Cipher.java:2074)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:31)
	at android.os.Handler.handleCallback(Handler.java:991)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8934)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)

  --- End of managed Java.Lang.Exception stack trace ---
android.security.KeyStoreException: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:835: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:385: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:90: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)) (public error code: 2 internal Keystore code: -26)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:428)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:79)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:118)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:179)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:624)
	at javax.crypto.Cipher.doFinal(Cipher.java:2074)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:31)
	at android.os.Handler.handleCallback(Handler.java:991)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8934)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)

   Exception_EndOfInnerExceptionStack
   at Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualObjectMethod(JniObjectReference , JniObjectReference , JniMethodInfo , JniArgumentValue* )
   at Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeNonvirtualObjectMethod(String , IJavaPeerable , JniArgumentValue* )
   at Javax.Crypto.Cipher.DoFinal(Byte[] )
   at keepass2android.BiometricDecryption.Decrypt(String encryted)
   at keepass2android.BiometricDecryption.DecryptStored(String prefKey)
   at keepass2android.PasswordActivity.OnBiometricAuthSucceeded()
  --- End of managed Java.Security.GeneralSecurityException stack trace ---
javax.crypto.IllegalBlockSizeException
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:634)
	at javax.crypto.Cipher.doFinal(Cipher.java:2074)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:31)
	at android.os.Handler.handleCallback(Handler.java:991)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8934)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)
Caused by: android.security.KeyStoreException: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:835: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:385: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:90: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)) (public error code: 2 internal Keystore code: -26)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:428)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:79)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:118)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:179)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:624)
	... 11 more

  --- End of managed Java.Security.GeneralSecurityException stack trace ---
javax.crypto.IllegalBlockSizeException
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:634)
	at javax.crypto.Cipher.doFinal(Cipher.java:2074)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:31)
	at android.os.Handler.handleCallback(Handler.java:991)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8934)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)
Caused by: android.security.KeyStoreException: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:835: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:385: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:90: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)) (public error code: 2 internal Keystore code: -26)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:428)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:79)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:118)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:179)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:624)
	... 11 more

12/05/2025 23:55:27:712 -- parsing autofillStructure...
12/05/2025 23:55:27:716 -- Parsing done
12/05/2025 23:55:27:717 -- cannot autofill
12/05/2025 23:55:29:838 -- CFS: Files in Sync
12/05/2025 23:55:29:888 -- Pre-loading database file completed
12/05/2025 23:55:34:340 -- PasswordActivity.OnPause 3
12/05/2025 23:55:34:399 -- AppSettingsActivity.OnCreate 4
12/05/2025 23:55:34:399 -- AppSettingsActivity:apptask= 4
12/05/2025 23:55:34:405 -- AppSettingsActivity.OnStart 4
12/05/2025 23:55:34:406 -- AppSettingsActivity.OnResume 4
12/05/2025 23:55:34:406 --  DB null 4
12/05/2025 23:55:34:921 -- PasswordActivity.OnStop 3
12/05/2025 23:55:36:809 -- AppSettingsActivity.OnPause 4
12/05/2025 23:55:37:391 -- AppSettingsActivity.OnStop 4
12/05/2025 23:55:39:824 -- AppSettingsActivity.OnStart 4
12/05/2025 23:55:39:825 -- AppSettingsActivity.OnResume 4
12/05/2025 23:55:39:825 --  DB null 4
12/05/2025 23:55:44:767 -- AppSettingsActivity.OnPause 4
12/05/2025 23:55:49:232 -- AppSettingsActivity.OnResume 4
12/05/2025 23:55:49:232 --  DB null 4
12/05/2025 23:55:50:494 -- AppSettingsActivity.OnPause 4
12/05/2025 23:55:50:982 -- AppSettingsActivity.OnStop 4
12/05/2025 23:56:06:512 -- parsing autofillStructure...
12/05/2025 23:56:06:513 -- Parsing done
12/05/2025 23:56:06:514 -- cannot autofill
12/05/2025 23:56:37:263 -- parsing autofillStructure...
12/05/2025 23:56:37:267 -- DAL verification failed for com.foxdebug.acodefree/localhost
12/05/2025 23:56:37:268 -- Parsing done
12/05/2025 23:56:37:268 -- cannot autofill
```

Same issue here on a Motorola Edge 30 Neo, Biometric unlock fails a couple of hours after enabling it.

<details><summary>Details</summary>
<p>


26/05/2025 17:38:43:178 -- DatabaseSettingsActivity.OnPause 9

26/05/2025 17:38:43:181 -- Timeout start
26/05/2025 17:38:43:190 -- GroupActivity.OnStart 7
26/05/2025 17:38:43:191 -- GroupActivity.OnResume 7
26/05/2025 17:38:43:191 --  DatabaseIsUnlocked=True 7
26/05/2025 17:38:43:191 -- Timeout cancel
26/05/2025 17:38:43:744 -- DatabaseSettingsActivity.OnStop 9
26/05/2025 17:38:43:747 -- DatabaseSettingsActivity.OnDestroyTrue 9
26/05/2025 17:38:46:75 -- Locking database
26/05/2025 17:38:46:115 -- OngoingNotificationsService.OnDestroy
26/05/2025 17:38:46:118 -- Finishing crc64c98c008c0cd742cb.GroupActivity due to database lock
26/05/2025 17:38:46:120 -- GroupActivity.OnPause 7
26/05/2025 17:38:46:131 -- SelectCurrentDbActivity.OnStart 5
26/05/2025 17:38:46:132 -- SelectCurrentDbActivity 5: OnActivityResult FirstUser/0
26/05/2025 17:38:46:132 -- TryGetFromActivityResult: no data
26/05/2025 17:38:46:133 -- SelectCurrentDbActivity.OnResume 5
26/05/2025 17:38:46:133 --  DB null 5
26/05/2025 17:38:46:157 -- SelectCurrentDbActivity.OnPause 5
26/05/2025 17:38:46:179 -- FileSelect.OnCreate
26/05/2025 17:38:46:202 -- FileSelect.OnStart
26/05/2025 17:38:46:231 -- PasswordActivity.OnCreate 10
26/05/2025 17:38:46:232 -- PasswordActivity:apptask= 10
26/05/2025 17:38:46:281 -- GetIocFromLaunchIntent()
26/05/2025 17:38:46:281 -- no keyprovider specified
26/05/2025 17:38:46:282 -- Reset keyfile
26/05/2025 17:38:46:294 -- PasswordActivity.OnStart 10
26/05/2025 17:38:46:295 -- FP: Create BiometricDecryption
26/05/2025 17:38:46:295 -- FP: Init for Dec
26/05/2025 17:38:46:298 -- alias: keepass2android.kp2a_ioc_2D93EED3628C8DD14F134423D3FFB6E18D72F83F16E1081571C3B23AC90E0535
26/05/2025 17:38:46:298 -- KS: end aliases
26/05/2025 17:38:46:318 -- Fingerprint: StartListening 
26/05/2025 17:38:46:322 -- PasswordActivity.OnResume 10
26/05/2025 17:38:46:322 --  DB null 10
26/05/2025 17:38:46:323 -- starting: True, Finishing: False, _performingLoad: False
26/05/2025 17:38:46:338 -- content://Klirre-sync.kdbx isCached = True
26/05/2025 17:38:46:339 -- Pre-loading database file starting
26/05/2025 17:38:46:363 -- content://Klirre-sync.kdbx isCached = True
26/05/2025 17:38:46:382 -- content://Klirre-sync.kdbx localVersionHash = 2332709F8CD81647FAB2A5BD2A6A387ADE49508BA139FC3B3B85EC2496D38D28
26/05/2025 17:38:46:412 -- PasswordModeSpinner item selected: 0
26/05/2025 17:38:46:413 -- content://Klirre-sync.kdbx baseVersionHash = 2332709F8CD81647FAB2A5BD2A6A387ADE49508BA139FC3B3B85EC2496D38D28
26/05/2025 17:38:46:413 -- CFS: OpenWhenNoLocalChanges
26/05/2025 17:38:46:415 -- CFS: hashing cached version
26/05/2025 17:38:46:438 -- content://Klirre-sync.kdbx isCached = True
26/05/2025 17:38:46:438 -- keepass2android.Io.OfflineModeException: Working offline.
   at keepass2android.Io.OfflineSwitchableFileStorage.AssertOnline()
   at keepass2android.Io.OfflineSwitchableFileStorage.OpenFileForRead(IOConnectionInfo ioc)
   at keepass2android.Io.CachingFileStorage.UpdateCacheFromRemote(IOConnectionInfo ioc, String cachedFilePath)
   at keepass2android.Io.CachingFileStorage.OpenFileForReadWhenNoLocalChanges(IOConnectionInfo ioc, String cachedFilePath)
   at keepass2android.Io.CachingFileStorage.OpenFileForRead(IOConnectionInfo ioc)
26/05/2025 17:38:46:444 -- Pre-loading database file completed
26/05/2025 17:38:46:958 -- GroupActivity.OnStop 7
26/05/2025 17:38:46:960 -- GroupActivity.OnDestroyTrue 7
26/05/2025 17:38:46:992 -- SelectCurrentDbActivity.OnStop 5
26/05/2025 17:38:46:993 -- FileSelect.OnStop
26/05/2025 17:38:47:17 -- FileSelect.OnDestroyTrue
26/05/2025 17:38:48:237 -- Fingerprint: StopListening  having prompt 
26/05/2025 17:38:48:239 -- PasswordActivity.OnPause 10
26/05/2025 17:38:48:298 -- PasswordActivity.OnStop 10
26/05/2025 17:38:49:484 -- PasswordActivity.OnStart 10
26/05/2025 17:38:49:485 -- PasswordActivity.OnResume 10
26/05/2025 17:38:49:485 --  DB null 10
26/05/2025 17:38:49:488 -- starting: True, Finishing: False, _performingLoad: False
26/05/2025 17:38:49:511 -- content://Klirre-sync.kdbx isCached = True
26/05/2025 17:38:49:624 -- parsing autofillStructure...
26/05/2025 17:38:49:625 -- Parsing done
26/05/2025 17:38:49:625 -- cannot autofill
26/05/2025 17:38:52:518 -- Fingerprint: StopListening  having prompt 
26/05/2025 17:38:52:520 -- PasswordActivity.OnPause 10
26/05/2025 17:38:52:526 -- PasswordActivity.OnStop 10
26/05/2025 17:38:52:604 -- SelectCurrentDbActivity.OnDestroyTrue 5
26/05/2025 17:38:52:610 -- PasswordActivity.OnDestroyTrue 10
26/05/2025 17:41:36:730 -- SelectCurrentDbActivity.OnDestroyTrue 1
26/05/2025 17:42:20:637 -- Creating application keepass2android.keepass2android. Version=206
26/05/2025 17:42:20:751 -- parsing autofillStructure...
26/05/2025 17:42:20:763 -- Parsing done
26/05/2025 17:42:20:765 -- cannot autofill
26/05/2025 17:49:11:182 -- parsing autofillStructure...
26/05/2025 17:49:11:183 -- Parsing done
26/05/2025 17:49:11:185 -- cannot autofill
26/05/2025 19:01:38:338 -- Creating application keepass2android.keepass2android. Version=206
26/05/2025 19:01:38:501 -- parsing autofillStructure...
26/05/2025 19:01:38:509 -- Parsing done
26/05/2025 19:01:38:518 -- AF: (query.IncompatiblePackageAndDomain == false)
26/05/2025 19:01:38:520 -- AF: BuildEntryDatasets
26/05/2025 19:01:38:529 -- AF: BuildEntryDatasets found 0 entries
26/05/2025 19:01:38:650 -- return autofill success
26/05/2025 19:01:39:330 -- ChooseForAutofillActivityBase.OnCreate
26/05/2025 19:01:39:370 -- ChooseForAutofillActivityBase: ExtraDisplayWarning = None
26/05/2025 19:01:39:519 -- SelectCurrentDbActivity.OnCreate 0
26/05/2025 19:01:39:520 -- SelectCurrentDbActivity:apptask= 0
26/05/2025 19:01:39:550 -- Loaded task keepass2android.SearchUrlTask
26/05/2025 19:01:39:551 -- Task in activity SelectCurrentDbActivity 0 changed to SearchUrlTask
26/05/2025 19:01:39:565 -- SelectCurrentDbActivity.OnStart 0
26/05/2025 19:01:39:568 -- SelectCurrentDbActivity.OnResume 0
26/05/2025 19:01:39:569 --  DB null 0
26/05/2025 19:01:39:584 -- SelectCurrentDbActivity.OnPause 0
26/05/2025 19:01:39:606 -- FileSelect.OnCreate
26/05/2025 19:01:39:705 -- FileSelect.OnStart
26/05/2025 19:01:39:856 -- SelectCurrentDbActivity.OnStop 0
26/05/2025 19:01:39:858 -- SelectCurrentDbActivity.OnDestroyFalse 0
26/05/2025 19:01:39:864 -- SelectCurrentDbActivity.OnCreate 1
26/05/2025 19:01:39:865 -- SelectCurrentDbActivity:apptask= 1
26/05/2025 19:01:39:871 -- Loaded task keepass2android.NullTask
26/05/2025 19:01:39:871 -- Task in activity SelectCurrentDbActivity 1 changed to NullTask
26/05/2025 19:01:39:873 -- SelectCurrentDbActivity.OnStart 1
26/05/2025 19:01:39:876 -- SelectCurrentDbActivity.OnResume 1
26/05/2025 19:01:39:876 --  DB null 1
26/05/2025 19:01:39:884 -- SelectCurrentDbActivity.OnPause 1
26/05/2025 19:01:39:889 -- ChooseForAutofillActivityBase.OnCreate
26/05/2025 19:01:39:890 -- ChooseForAutofillActivityBase: ExtraDisplayWarning = None
26/05/2025 19:01:39:922 -- SelectCurrentDbActivity.OnCreate 2
26/05/2025 19:01:39:922 -- SelectCurrentDbActivity:apptask= 2
26/05/2025 19:01:39:927 -- Loaded task keepass2android.SearchUrlTask
26/05/2025 19:01:39:927 -- Task in activity SelectCurrentDbActivity 2 changed to SearchUrlTask
26/05/2025 19:01:39:929 -- SelectCurrentDbActivity.OnStart 2
26/05/2025 19:01:39:931 -- SelectCurrentDbActivity.OnResume 2
26/05/2025 19:01:39:931 --  DB null 2
26/05/2025 19:01:39:940 -- SelectCurrentDbActivity.OnPause 2
26/05/2025 19:01:39:954 -- FileSelect.OnCreate
26/05/2025 19:01:39:980 -- FileSelect.OnStart
26/05/2025 19:01:40:12 -- PasswordActivity.OnCreate 3
26/05/2025 19:01:40:13 -- PasswordActivity:apptask= 3
26/05/2025 19:01:40:142 -- GetIocFromLaunchIntent()
26/05/2025 19:01:40:142 -- no keyprovider specified
26/05/2025 19:01:40:144 -- Reset keyfile
26/05/2025 19:01:40:576 -- PasswordActivity.OnStart 3
26/05/2025 19:01:40:584 -- FP: Create BiometricDecryption
26/05/2025 19:01:40:593 -- FP: Init for Dec
26/05/2025 19:01:40:597 -- alias: keepass2android.kp2a_ioc_2D93EED3628C8DD14F134423D3FFB6E18D72F83F16E1081571C3B23AC90E0535
26/05/2025 19:01:40:598 -- KS: end aliases
26/05/2025 19:01:40:647 -- Fingerprint: StartListening 
26/05/2025 19:01:40:699 -- PasswordActivity.OnNewIntent 3
26/05/2025 19:01:40:701 -- GetIocFromLaunchIntent()
26/05/2025 19:01:40:701 -- no keyprovider specified
26/05/2025 19:01:40:701 -- Reset keyfile
26/05/2025 19:01:40:718 -- PasswordActivity.OnStart 3
26/05/2025 19:01:40:719 -- FP: Create BiometricDecryption
26/05/2025 19:01:40:720 -- FP: Init for Dec
26/05/2025 19:01:40:724 -- alias: keepass2android.kp2a_ioc_2D93EED3628C8DD14F134423D3FFB6E18D72F83F16E1081571C3B23AC90E0535
26/05/2025 19:01:40:724 -- KS: end aliases
26/05/2025 19:01:40:793 -- Fingerprint: StartListening 
26/05/2025 19:01:40:797 -- PasswordActivity.OnResume 3
26/05/2025 19:01:40:798 --  DB null 3
26/05/2025 19:01:40:803 -- starting: True, Finishing: False, _performingLoad: False
26/05/2025 19:01:40:835 -- content://Klirre-sync.kdbx isCached = True
26/05/2025 19:01:40:887 -- Pre-loading database file starting
26/05/2025 19:01:40:917 -- content://Klirre-sync.kdbx isCached = True
26/05/2025 19:01:40:938 -- content://Klirre-sync.kdbx localVersionHash = 2332709F8CD81647FAB2A5BD2A6A387ADE49508BA139FC3B3B85EC2496D38D28
26/05/2025 19:01:40:953 -- content://Klirre-sync.kdbx baseVersionHash = 2332709F8CD81647FAB2A5BD2A6A387ADE49508BA139FC3B3B85EC2496D38D28
26/05/2025 19:01:40:953 -- CFS: OpenWhenNoLocalChanges
26/05/2025 19:01:40:954 -- CFS: hashing cached version
26/05/2025 19:01:40:991 -- content://Klirre-sync.kdbx isCached = True
26/05/2025 19:01:40:998 -- keepass2android.Io.OfflineModeException: Working offline.
   at keepass2android.Io.OfflineSwitchableFileStorage.AssertOnline()
   at keepass2android.Io.OfflineSwitchableFileStorage.OpenFileForRead(IOConnectionInfo ioc)
   at keepass2android.Io.CachingFileStorage.UpdateCacheFromRemote(IOConnectionInfo ioc, String cachedFilePath)
   at keepass2android.Io.CachingFileStorage.OpenFileForReadWhenNoLocalChanges(IOConnectionInfo ioc, String cachedFilePath)
   at keepass2android.Io.CachingFileStorage.OpenFileForRead(IOConnectionInfo ioc)
26/05/2025 19:01:41:96 -- PasswordModeSpinner item selected: 0
26/05/2025 19:01:41:96 -- PasswordModeSpinner item selected: 0
26/05/2025 19:01:41:108 -- Pre-loading database file completed
26/05/2025 19:01:41:718 -- SelectCurrentDbActivity.OnStop 1
26/05/2025 19:01:41:719 -- FileSelect.OnStop
26/05/2025 19:01:41:742 -- FileSelect.OnDestroyTrue
26/05/2025 19:01:41:744 -- SelectCurrentDbActivity.OnStop 2
26/05/2025 19:01:41:745 -- SelectCurrentDbActivity.OnDestroyTrue 2
26/05/2025 19:01:41:746 -- FileSelect.OnStop
26/05/2025 19:01:41:783 -- FileSelect.OnDestroyTrue
26/05/2025 19:01:42:812 -- FP: Decrypting 
26/05/2025 19:01:42:826 -- GeneralSecurityException in DecryptStored
26/05/2025 19:01:42:839 -- Java.Security.GeneralSecurityException: Exception_WasThrown, Java.Security.GeneralSecurityException
 ---> Java.Lang.Exception: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:836: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:383: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:99: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED))

  --- End of managed Java.Lang.Exception stack trace ---
android.security.KeyStoreException: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:836: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:383: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:99: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)) (public error code: 2 internal Keystore code: -26)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:386)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:78)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:115)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:179)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:618)
	at javax.crypto.Cipher.doFinal(Cipher.java:2074)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:31)
	at android.os.Handler.handleCallback(Handler.java:1000)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loopOnce(Looper.java:242)
	at android.os.Looper.loop(Looper.java:362)
	at android.app.ActivityThread.main(ActivityThread.java:8393)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:992)

  --- End of managed Java.Lang.Exception stack trace ---
android.security.KeyStoreException: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:836: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:383: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:99: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)) (public error code: 2 internal Keystore code: -26)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:386)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:78)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:115)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:179)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:618)
	at javax.crypto.Cipher.doFinal(Cipher.java:2074)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:31)
	at android.os.Handler.handleCallback(Handler.java:1000)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loopOnce(Looper.java:242)
	at android.os.Looper.loop(Looper.java:362)
	at android.app.ActivityThread.main(ActivityThread.java:8393)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:992)

   Exception_EndOfInnerExceptionStack
   at Java.Interop.JniEnvironment.InstanceMethods.CallNonvirtualObjectMethod(JniObjectReference , JniObjectReference , JniMethodInfo , JniArgumentValue* )
   at Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeNonvirtualObjectMethod(String , IJavaPeerable , JniArgumentValue* )
   at Javax.Crypto.Cipher.DoFinal(Byte[] )
   at keepass2android.BiometricDecryption.Decrypt(String encryted)
   at keepass2android.BiometricDecryption.DecryptStored(String prefKey)
   at keepass2android.PasswordActivity.OnBiometricAuthSucceeded()
  --- End of managed Java.Security.GeneralSecurityException stack trace ---
javax.crypto.IllegalBlockSizeException
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:628)
	at javax.crypto.Cipher.doFinal(Cipher.java:2074)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:31)
	at android.os.Handler.handleCallback(Handler.java:1000)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loopOnce(Looper.java:242)
	at android.os.Looper.loop(Looper.java:362)
	at android.app.ActivityThread.main(ActivityThread.java:8393)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:992)
Caused by: android.security.KeyStoreException: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:836: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:383: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:99: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)) (public error code: 2 internal Keystore code: -26)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:386)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:78)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:115)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:179)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:618)
	... 11 more

  --- End of managed Java.Security.GeneralSecurityException stack trace ---
javax.crypto.IllegalBlockSizeException
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:628)
	at javax.crypto.Cipher.doFinal(Cipher.java:2074)
	at mono.java.lang.RunnableImplementor.n_run(Native Method)
	at mono.java.lang.RunnableImplementor.run(RunnableImplementor.java:31)
	at android.os.Handler.handleCallback(Handler.java:1000)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loopOnce(Looper.java:242)
	at android.os.Looper.loop(Looper.java:362)
	at android.app.ActivityThread.main(ActivityThread.java:8393)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:992)
Caused by: android.security.KeyStoreException: Key user not authenticated (internal Keystore code: -26 message: system/security/keystore2/src/operation.rs:836: KeystoreOperation::update

Caused by:
    0: system/security/keystore2/src/operation.rs:383: Trying to get auth tokens.
    1: In AuthInfo::get_auth_tokens.
    2: system/security/keystore2/src/enforcements.rs:99: No operation auth token received.
    3: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)) (public error code: 2 internal Keystore code: -26)
	at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:386)
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:78)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:115)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:179)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:618)
	... 11 more

26/05/2025 19:01:42:955 -- parsing autofillStructure...
26/05/2025 19:01:42:968 -- Parsing done
26/05/2025 19:01:42:969 -- cannot autofill
26/05/2025 19:01:50:808 -- PasswordActivity.OnPause 3
26/05/2025 19:01:50:892 -- AppSettingsActivity.OnCreate 4
26/05/2025 19:01:50:893 -- AppSettingsActivity:apptask= 4
26/05/2025 19:01:50:907 -- AppSettingsActivity.OnStart 4
26/05/2025 19:01:50:908 -- AppSettingsActivity.OnResume 4
26/05/2025 19:01:50:909 --  DB null 4
26/05/2025 19:01:51:391 -- PasswordActivity.OnStop 3
26/05/2025 19:01:54:506 -- AppSettingsActivity.OnPause 4
26/05/2025 19:01:58:62 -- AppSettingsActivity.OnResume 4
26/05/2025 19:01:58:63 --  DB null 4
26/05/2025 19:02:00:369 -- PasswordActivity.OnStart 3
26/05/2025 19:02:00:370 -- PasswordActivity.OnResume 3
26/05/2025 19:02:00:370 --  DB null 3
26/05/2025 19:02:00:371 -- starting: True, Finishing: False, _performingLoad: False
26/05/2025 19:02:00:399 -- content://Klirre-sync.kdbx isCached = True
26/05/2025 19:02:00:421 -- AppSettingsActivity.OnPause 4
26/05/2025 19:02:00:461 -- AppSettingsActivity.OnStop 4
26/05/2025 19:02:02:503 -- PasswordActivity.OnPause 3
26/05/2025 19:02:02:549 -- PasswordActivity.OnStop 3
26/05/2025 19:02:18:844 -- parsing autofillStructure...
26/05/2025 19:02:18:845 -- Parsing done
26/05/2025 19:02:18:845 -- AF: (query.IncompatiblePackageAndDomain == false)
26/05/2025 19:02:18:846 -- AF: BuildEntryDatasets
26/05/2025 19:02:18:846 -- AF: BuildEntryDatasets found 0 entries
26/05/2025 19:02:18:850 -- return autofill success
26/05/2025 19:04:13:139 -- parsing autofillStructure...
26/05/2025 19:04:13:139 -- Parsing done
26/05/2025 19:04:13:140 -- AF: (query.IncompatiblePackageAndDomain == false)
26/05/2025 19:04:13:140 -- AF: BuildEntryDatasets
26/05/2025 19:04:13:141 -- AF: BuildEntryDatasets found 0 entries
26/05/2025 19:04:13:144 -- return autofill success
26/05/2025 19:05:03:532 -- AppSettingsActivity.OnStart 4
26/05/2025 19:05:03:534 -- AppSettingsActivity.OnResume 4
26/05/2025 19:05:03:534 --  DB null 4


</p>
</details>

--- a/src/keepass2android-app/PasswordActivity.cs
+++ b/src/keepass2android-app/PasswordActivity.cs
@@ -131,7 +131,8 @@ namespace keepass2android
 		ISharedPreferences _prefs;
 
 		private bool _starting;
-		private OtpInfo _otpInfo;
+        private bool _resumeCompleted;
+        private OtpInfo _otpInfo;
 		private IOConnectionInfo _otpAuxIoc;
         private ChallengeInfo _chalInfo;
         private byte[] _challengeSecret;
@@ -1569,7 +1570,9 @@ namespace keepass2android
 		}
 
         private bool hasRequestedKeyboardActivation = false;
-		protected override void OnStart()
+        
+
+        protected override void OnStart()
 		{
 			base.OnStart();
 			_starting = true;
@@ -1644,60 +1647,65 @@ namespace keepass2android
             if (intent != null)
             {
 				if (intent.HasExtra(Intents.OtpExtraKey))
-                {
-                    string otp = intent.GetStringExtra(Intents.OtpExtraKey);
-                    _keepPasswordInOnResume = true;
-                    if (KeyProviderTypes.Contains(KeyProviders.Otp))
+				{
+					string otp = intent.GetStringExtra(Intents.OtpExtraKey);
+					_keepPasswordInOnResume = true;
+					if (KeyProviderTypes.Contains(KeyProviders.Otp))
+					{
+
+						if (_otpInfo == null)
+						{
+							//Entering OTPs not yet initialized:
+							_pendingOtps.Add(otp);
+							UpdateKeyProviderUiState();
+						}
+						else
+						{
+							//Entering OTPs is initialized. Write OTP into first empty field:
+							bool foundEmptyField = false;
+							foreach (int otpId in _otpTextViewIds)
+							{
+								EditText otpEdit = FindViewById<EditText>(otpId);
+								if ((otpEdit.Visibility == ViewStates.Visible) && String.IsNullOrEmpty(otpEdit.Text))
+								{
+									otpEdit.Text = otp;
+									foundEmptyField = true;
+									break;
+								}
+							}
+							//did we find a field?
+							if (!foundEmptyField)
+							{
+								App.Kp2a.ShowMessage(this, GetString(Resource.String.otp_discarded_no_space), MessageSeverity.Error);
+							}
+						}
+
+						Spinner passwordModeSpinner = FindViewById<Spinner>(Resource.Id.password_mode_spinner);
+						if (passwordModeSpinner.SelectedItemPosition != (int)KeyProviders.Otp)
+						{
+							passwordModeSpinner.SetSelection((int)KeyProviders.Otp);
+						}
+					}
+					else
+					{
+						//assume the key should be used as static password
+						FindViewById<EditText>(Resource.Id.password_edit).Text += otp;
+					}
+				}
+				else
+				{
+                    // if the activity is launched twice and the first initialization hasn't even finished, we cannot 
+                    // reset the state and re-initialize the activity.
+					// This can happen with autofill in some cases (#2869)
+                    if (_resumeCompleted)
                     {
-
-                        if (_otpInfo == null)
-                        {
-                            //Entering OTPs not yet initialized:
-                            _pendingOtps.Add(otp);
-                            UpdateKeyProviderUiState();
-                        }
-                        else
-                        {
-                            //Entering OTPs is initialized. Write OTP into first empty field:
-                            bool foundEmptyField = false;
-                            foreach (int otpId in _otpTextViewIds)
-                            {
-                                EditText otpEdit = FindViewById<EditText>(otpId);
-                                if ((otpEdit.Visibility == ViewStates.Visible) && String.IsNullOrEmpty(otpEdit.Text))
-                                {
-                                    otpEdit.Text = otp;
-                                    foundEmptyField = true;
-                                    break;
-                                }
-                            }
-                            //did we find a field?
-                            if (!foundEmptyField)
-                            {
-                                App.Kp2a.ShowMessage(this, GetString(Resource.String.otp_discarded_no_space),  MessageSeverity.Error);
-                            }
-                        }
-
-                        Spinner passwordModeSpinner = FindViewById<Spinner>(Resource.Id.password_mode_spinner);
-                        if (passwordModeSpinner.SelectedItemPosition != (int)KeyProviders.Otp)
-                        {
-                            passwordModeSpinner.SetSelection((int)KeyProviders.Otp);
-                        }
+                        ResetState();
+                        GetIocFromLaunchIntent(intent);
+                        InitializeAfterSetIoc();
+                        OnStart();
                     }
-                    else
-                    {
-                        //assume the key should be used as static password
-                        FindViewById<EditText>(Resource.Id.password_edit).Text += otp;
-                    }
-
-
-                }
-                else
-                {
-                    ResetState();
-                    GetIocFromLaunchIntent(intent);
-					InitializeAfterSetIoc();
-                    OnStart();
-                }
+					
+				}
 			}
 	
 		}
@@ -1736,148 +1744,150 @@ namespace keepass2android
 
         protected override View? SnackbarAnchorView => FindViewById(Resource.Id.main_content);
 
-        protected override void OnResume()
-        {
-            base.OnResume();
-            
-            _activityDesign.ReapplyTheme();
+		protected override void OnResume()
+		{
+			base.OnResume();
 
-            Kp2aLog.Log("starting: " + _starting + ", Finishing: " + IsFinishing + ", _performingLoad: " +
-                        _performingLoad);
+			_activityDesign.ReapplyTheme();
 
-            CheckBox cbOfflineMode = (CheckBox)FindViewById(Resource.Id.work_offline);
-            App.Kp2a.OfflineMode =
-                cbOfflineMode.Checked =
-                    App.Kp2a
-                        .OfflineModePreference; //this won't overwrite new user settings because every change is directly saved in settings
-            LinearLayout offlineModeContainer = FindViewById<LinearLayout>(Resource.Id.work_offline_container);
-            var cachingFileStorage = App.Kp2a.GetFileStorage(_ioConnection) as CachingFileStorage;
-            if ((cachingFileStorage != null) && cachingFileStorage.IsCached(_ioConnection))
-            {
-                offlineModeContainer.Visibility = ViewStates.Visible;
-            }
-            else
-            {
-                offlineModeContainer.Visibility = ViewStates.Gone;
-                App.Kp2a.OfflineMode = false;
-            }
+			Kp2aLog.Log("starting: " + _starting + ", Finishing: " + IsFinishing + ", _performingLoad: " +
+						_performingLoad);
 
-
-
-
-            View killButton = FindViewById(Resource.Id.kill_app);
-            if (PreferenceManager.GetDefaultSharedPreferences(this)
-                .GetBoolean(GetString(Resource.String.show_kill_app_key), false))
-            {
-                killButton.Click += (sender, args) =>
-                {
-                    _killOnDestroy = true;
-                    SetResult(Result.Canceled);
-                    Finish();
-
-                };
-                killButton.Visibility = ViewStates.Visible;
-
-            }
-            else
-            {
-                killButton.Visibility = ViewStates.Gone;
-            }
-
-            TryGetOtpFromClipboard();
-
-            if (!_keepPasswordInOnResume)
-            {
-                if (
-                    _lastOnPauseTime <
-                    DateTime.Now -
-                    TimeSpan.FromSeconds(
-                        5) //only clear when user left the app for more than 5 seconds (allows to use Yubiclip, also allows to switch shortly to another app)
-                    &&
-                    PreferenceManager.GetDefaultSharedPreferences(this)
-                        .GetBoolean(GetString(Resource.String.ClearPasswordOnLeave_key), true))
-                {
-                    ClearEnteredPassword();
-                }
-
-            }
-
-
-
-            _keepPasswordInOnResume = false;
-
-            MakePasswordMaskedOrVisible();
-
-            UpdateOkButtonState();
-
-            if (KeyProviderTypes.Contains(KeyProviders.Challenge))
-            {
-                FindViewById(Resource.Id.otpInitView).Visibility =
-                    _challengeSecret == null ? ViewStates.Visible : ViewStates.Gone;
-            }
-
-            //use !IsFinishing to make sure we're not starting another activity when we're already finishing (e.g. due to TaskComplete in OnActivityResult)
-            //use !performingLoad to make sure we're not already loading the database (after ActivityResult from File-Prepare-Activity; this would cause _loadDbFileTask to exist when we reload later!)
-            if ( !IsFinishing && !_performingLoad)  
+			CheckBox cbOfflineMode = (CheckBox)FindViewById(Resource.Id.work_offline);
+			App.Kp2a.OfflineMode =
+				cbOfflineMode.Checked =
+					App.Kp2a
+						.OfflineModePreference; //this won't overwrite new user settings because every change is directly saved in settings
+			LinearLayout offlineModeContainer = FindViewById<LinearLayout>(Resource.Id.work_offline_container);
+			var cachingFileStorage = App.Kp2a.GetFileStorage(_ioConnection) as CachingFileStorage;
+			if ((cachingFileStorage != null) && cachingFileStorage.IsCached(_ioConnection))
 			{
-				
-				
-			    // OnResume is run every time the activity comes to the foreground. This code should only run when the activity is started (OnStart), but must
-				// be run in OnResume rather than OnStart so that it always occurrs after OnActivityResult (when re-creating a killed activity, OnStart occurs before OnActivityResult)
-			    if (_starting)
-			    {
+				offlineModeContainer.Visibility = ViewStates.Visible;
+			}
+			else
+			{
+				offlineModeContainer.Visibility = ViewStates.Gone;
+				App.Kp2a.OfflineMode = false;
+			}
 
-			        _starting = false;
 
-			        //database not yet loaded.
 
-			        //check if pre-loading is enabled but wasn't started yet:
-			        if (_loadDbFileTask == null &&
-			            _prefs.GetBoolean(GetString(Resource.String.PreloadDatabaseEnabled_key), true))
-			        {
-			            // Create task to kick off file loading while the user enters the password
-			            _loadDbFileTask = Task.Factory.StartNew(PreloadDbFile);
-			            _loadDbTaskOffline = App.Kp2a.OfflineMode;
-			        }
-			    }
+
+			View killButton = FindViewById(Resource.Id.kill_app);
+			if (PreferenceManager.GetDefaultSharedPreferences(this)
+				.GetBoolean(GetString(Resource.String.show_kill_app_key), false))
+			{
+				killButton.Click += (sender, args) =>
+				{
+					_killOnDestroy = true;
+					SetResult(Result.Canceled);
+					Finish();
+
+				};
+				killButton.Visibility = ViewStates.Visible;
+
+			}
+			else
+			{
+				killButton.Visibility = ViewStates.Gone;
+			}
+
+			TryGetOtpFromClipboard();
+
+			if (!_keepPasswordInOnResume)
+			{
+				if (
+					_lastOnPauseTime <
+					DateTime.Now -
+					TimeSpan.FromSeconds(
+						5) //only clear when user left the app for more than 5 seconds (allows to use Yubiclip, also allows to switch shortly to another app)
+					&&
+					PreferenceManager.GetDefaultSharedPreferences(this)
+						.GetBoolean(GetString(Resource.String.ClearPasswordOnLeave_key), true))
+				{
+					ClearEnteredPassword();
+				}
 
 			}
 
-		    if (compositeKeyForImmediateLoad != null)
-		    {
-		        //reload the database (without most other stuff performed in PerformLoadDatabase.
-		        // We're assuming that the db file (and if appropriate also the key file) are still available 
-		        // and there's no need to re-init the file storage. if it is, loading will fail and the user has 
-		        // to retry with typing the full password, but that's intended to avoid showing the password to a 
-		        // a potentially unauthorized user (feature request https://keepass2android.codeplex.com/workitem/274)
-		        Handler handler = new Handler();
-		        OnFinish onFinish = new AfterLoad(handler, this, _ioConnection);
-		        _performingLoad = true;
-                LoadDb task = new LoadDb(this, App.Kp2a, _ioConnection, _loadDbFileTask, compositeKeyForImmediateLoad, GetKeyProviderString(),
-		            onFinish, false, _makeCurrent);
-		        _loadDbFileTask = null; // prevent accidental re-use
-		        new ProgressTask(App.Kp2a, this, task).Run();
-		        compositeKeyForImmediateLoad = null; //don't reuse or keep in memory
 
-		    }
-		    else
-		    {
-		        bool showKeyboard = true;
 
-		       
-		        EditText pwd = (EditText) FindViewById(Resource.Id.password_edit);
-		        pwd.PostDelayed(() =>
-		        {
-		            InputMethodManager keyboard = (InputMethodManager) GetSystemService(InputMethodService);
-                    if (showKeyboard)
-                    {
-                        pwd.RequestFocus();
-                        keyboard.ShowSoftInput(pwd, 0);
-                    }
-                    else
-		                keyboard.HideSoftInputFromWindow(pwd.WindowToken, HideSoftInputFlags.ImplicitOnly);
-		        }, 50);
-		    }
+			_keepPasswordInOnResume = false;
+
+			MakePasswordMaskedOrVisible();
+
+			UpdateOkButtonState();
+
+			if (KeyProviderTypes.Contains(KeyProviders.Challenge))
+			{
+				FindViewById(Resource.Id.otpInitView).Visibility =
+					_challengeSecret == null ? ViewStates.Visible : ViewStates.Gone;
+			}
+
+			//use !IsFinishing to make sure we're not starting another activity when we're already finishing (e.g. due to TaskComplete in OnActivityResult)
+			//use !performingLoad to make sure we're not already loading the database (after ActivityResult from File-Prepare-Activity; this would cause _loadDbFileTask to exist when we reload later!)
+			if (!IsFinishing && !_performingLoad)
+			{
+
+
+				// OnResume is run every time the activity comes to the foreground. This code should only run when the activity is started (OnStart), but must
+				// be run in OnResume rather than OnStart so that it always occurrs after OnActivityResult (when re-creating a killed activity, OnStart occurs before OnActivityResult)
+				if (_starting)
+				{
+
+					_starting = false;
+
+					//database not yet loaded.
+
+					//check if pre-loading is enabled but wasn't started yet:
+					if (_loadDbFileTask == null &&
+						_prefs.GetBoolean(GetString(Resource.String.PreloadDatabaseEnabled_key), true))
+					{
+						// Create task to kick off file loading while the user enters the password
+						_loadDbFileTask = Task.Factory.StartNew(PreloadDbFile);
+						_loadDbTaskOffline = App.Kp2a.OfflineMode;
+					}
+				}
+
+			}
+
+			if (compositeKeyForImmediateLoad != null)
+			{
+				//reload the database (without most other stuff performed in PerformLoadDatabase.
+				// We're assuming that the db file (and if appropriate also the key file) are still available 
+				// and there's no need to re-init the file storage. if it is, loading will fail and the user has 
+				// to retry with typing the full password, but that's intended to avoid showing the password to a 
+				// a potentially unauthorized user (feature request https://keepass2android.codeplex.com/workitem/274)
+				Handler handler = new Handler();
+				OnFinish onFinish = new AfterLoad(handler, this, _ioConnection);
+				_performingLoad = true;
+				LoadDb task = new LoadDb(this, App.Kp2a, _ioConnection, _loadDbFileTask, compositeKeyForImmediateLoad, GetKeyProviderString(),
+					onFinish, false, _makeCurrent);
+				_loadDbFileTask = null; // prevent accidental re-use
+				new ProgressTask(App.Kp2a, this, task).Run();
+				compositeKeyForImmediateLoad = null; //don't reuse or keep in memory
+
+			}
+			else
+			{
+				bool showKeyboard = true;
+
+
+				EditText pwd = (EditText)FindViewById(Resource.Id.password_edit);
+				pwd.PostDelayed(() =>
+				{
+					InputMethodManager keyboard = (InputMethodManager)GetSystemService(InputMethodService);
+					if (showKeyboard)
+					{
+						pwd.RequestFocus();
+						keyboard.ShowSoftInput(pwd, 0);
+					}
+					else
+						keyboard.HideSoftInputFromWindow(pwd.WindowToken, HideSoftInputFlags.ImplicitOnly);
+				}, 50);
+			}
+
+			_resumeCompleted = true;
         }
 
 	    private void TryGetOtpFromClipboard()

